diopter/compiler.py

```diff
@@ -138,8 +140,9 @@ class SourceProgram:
             flags, prefixed with a dash ("-") that will be passed to the compiler
     """
 
-    code: str
     language: Language
+    code: str | None = None
+    filename: Path | None = None
```

Comment:
Instead of having these two optional fields, I think it would be better to have to different types, one that is file based and another that is similar to `SourceProgram`. 

They need to have some common interface, e.g., they both have a method that returns a pathlike object (the `SourceProgram` creates a tempfile and returns that, and the other class returns the filename), and a method that returns relevant flags. 

The interface would then be cleaner and there's no need to check everywhere for `Nones`